src/Microsoft.AspNetCore.Mvc.TagHelpers/TagHelperOutputExtensions.cs

```diff
@@ -154,6 +156,230 @@ public static void MergeAttributes(this TagHelperOutput tagHelperOutput, TagBuil
+            this TagHelperOutput tagHelperOutput,
+            string classValue,
+            HtmlEncoder htmlEncoder)
+        {
+            if (tagHelperOutput == null)
+            {
+                throw new ArgumentNullException(nameof(tagHelperOutput));
+            }
+
+            if (string.IsNullOrEmpty(classValue))
+            {
+                return;
+            }
+
+            var encodedSpaceChars = SpaceChars.Where(x => !x.Equals('\u0020')).Select(x => htmlEncoder.Encode(x.ToString())).ToArray();
+
+            if (SpaceChars.Any(classValue.Contains) || encodedSpaceChars.Any(classValue.Contains))
+            {
+                throw new ArgumentException(Resources.ArgumentCannotContainHtmlSpace, nameof(classValue));
+            }
+
+            if (!tagHelperOutput.Attributes.TryGetAttribute("class", out TagHelperAttribute classAttribute))
+            {
+                tagHelperOutput.Attributes.Add("class", classValue);
+            }
+            else
+            {
+                var currentClassValue = ExtractClassValue(classAttribute, htmlEncoder);
+
+                var encodedClassValue = htmlEncoder.Encode(classValue);
+
+                if (currentClassValue.Equals(encodedClassValue))
+                {
+                    return;
+                }
+
+                var containsSpace = SpaceChars.Any(currentClassValue.Contains);
+
+                var containsEncodedSpace = encodedSpaceChars.Any(currentClassValue.Contains);
+
+                if (containsSpace && !containsEncodedSpace)
+                {
+                    var arrayOfClasses = currentClassValue.Split(SpaceChars);
+
+                    if (arrayOfClasses.Contains(encodedClassValue))
+                    {
+                        return;
+                    }
+                }
+                else if (!containsSpace && containsEncodedSpace)
+                {
+                    var arrayOfClasses = currentClassValue.Split(encodedSpaceChars, StringSplitOptions.None);
+
+                    if (arrayOfClasses.Contains(encodedClassValue))
+                    {
+                        return;
+                    }
+                }
+                else if (containsSpace)
+                {
+                    var listOfClasses = currentClassValue.Split(SpaceChars).ToList();
+                    var classesContainingEncodedSpace = listOfClasses.Where(x => encodedSpaceChars.Any(x.Contains));
+                    var listOfSubClasses = new List<string>();
+
+                    foreach (var classContainingEncodedSpace in classesContainingEncodedSpace)
+                    {
+                        var subClasses = classContainingEncodedSpace.Split(encodedSpaceChars, StringSplitOptions.None);
+                        listOfSubClasses.AddRange(subClasses);
+                    }
+
+                    listOfClasses.RemoveAll(x => encodedSpaceChars.Any(x.Contains));
+                    listOfClasses.AddRange(listOfSubClasses);
```

Comment:
This reorders the classes and that seems unecessary. Would the following work?
``` c#
var arrayOfClasses = currentClassValue.Split(SpaceChars, StringSplitOptions.Y)
    .SelectMany(perhapsEncoded => perhapsEncoded.Split(encodedSpaceChars, StringSplitOptions.Y))
    .ToArray();
```

If yes, can move the `Contains(encodedClassValue)` checks out of these blocks.

Reply:
I did think of doing it like this and remove the whole `containsSpace`/`containsEncodedSpace` (lines 203-244). Would that be better/efficient than checking if those chars actually exist before we split though?

Reply:
It does work btw for both AddClass and RemoveClass. RemoveClass need it to be a List<string> for the RemoveAll method but other than that it works like a charm.

Reply:
I'm fine with keeping the `containsSpace` / `containsEncodedSpace` split and doing the `SelectMany(...)` bit only in the "both" case. I'm also fine with always doing everything. It's probably slightly more efficient to use `SelectMany(...)` only when it's required but simpler code leans me toward doing it all the time.

Reply:
It looks so much cleaner. Will check it in now.